File: d09/d09.exs
solve = fn first? ->
  File.stream!("input.txt")
  |> Stream.map(fn ln -> 
    String.split(ln)
    |> Enum.map(&String.to_integer/1)
    |> Stream.unfold(fn l ->
      case Enum.all?(l, & &1 == 0) do
        true -> nil
        false -> {Enum.at(l, first? && 0 || -1), Enum.zip_with(tl(l), l, &-/2)}
      end
    end)
    |> then(fn ends ->
      case first? do
        true -> ends |> Enum.reverse() |> Enum.reduce(&-/2)
        false -> ends |> Enum.sum()
      end
    end)
  end)
  |> Enum.sum()
end

IO.inspect(solve.(false), label: "Star 1")
IO.inspect(solve.(true), label: "Star 2")

File: d09/d09.py
from operator import sub, add
l = [[*map(int,l.split())] for l in open(0)]

def f(l, pos, op):
    return op(l[pos], f([*map(sub, l[1:], l)], pos, op)) if l else 0

for pos, op in (-1, add), (0, sub): 
    print(sum(map(lambda l: f(l, pos, op), l)))
